File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/AcceptRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import java.util.Optional;
import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.constant.MessageType;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.AcceptRequest;
import net.prostars.messagesystem.dto.websocket.outbound.AcceptNotification;
import net.prostars.messagesystem.dto.websocket.outbound.AcceptResponse;
import net.prostars.messagesystem.dto.websocket.outbound.ErrorResponse;
import net.prostars.messagesystem.service.ClientNotificationService;
import net.prostars.messagesystem.service.UserConnectionService;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class AcceptRequestHandler implements BaseRequestHandler<AcceptRequest> {

  private final UserConnectionService userConnectionService;
  private final ClientNotificationService clientNotificationService;

  public AcceptRequestHandler(
      UserConnectionService userConnectionService,
      ClientNotificationService clientNotificationService) {
    this.userConnectionService = userConnectionService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, AcceptRequest request) {
    UserId acceptorUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());
    Pair<Optional<UserId>, String> result =
        userConnectionService.accept(acceptorUserId, request.getUsername());
    result
        .getFirst()
        .ifPresentOrElse(
            inviterUserId -> {
              clientNotificationService.sendMessage(
                  senderSession, acceptorUserId, new AcceptResponse(request.getUsername()));
              String acceptorUsername = result.getSecond();
              clientNotificationService.sendMessage(
                  inviterUserId, new AcceptNotification(acceptorUsername));
            },
            () -> {
              String errorMessage = result.getSecond();
              clientNotificationService.sendMessage(
                  senderSession,
                  acceptorUserId,
                  new ErrorResponse(MessageType.ACCEPT_REQUEST, errorMessage));
            });
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/AcceptRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.constant.MessageType
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.AcceptRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.AcceptNotification
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.AcceptResponse
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.ErrorResponse
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import com.yscorp.withpush.messagesystem.service.UserConnectionService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
class AcceptRequestHandler(
    private val userConnectionService: UserConnectionService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<AcceptRequest> {

    override fun handleRequest(senderSession: WebSocketSession, request: AcceptRequest) {
        val acceptorUserId: UserId = senderSession.attributes[IdKey.USER_ID.value] as UserId
        val (userId, errorMessage) = userConnectionService.accept(acceptorUserId, request.username)

        if (userId == null) {
            val errorMessage = errorMessage

            clientNotificationService.sendMessage(
                senderSession,
                acceptorUserId,
                ErrorResponse(MessageType.ACCEPT_REQUEST, errorMessage)
            )
        } else {

            clientNotificationService.sendMessage(
                senderSession, acceptorUserId, AcceptResponse(request.username)
            )
            clientNotificationService.sendMessage(
                userId, AcceptNotification(errorMessage)
            )
        }
    }
}
